In [1]:
# Initial imports
import pandas as pd
from sklearn import tree
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
%matplotlib inline

## **Loading and Preprocessing the Data**

In [2]:
# Loading data
url = 'https://raw.githubusercontent.com/bernardtse/project_4_machine_learning/main/resources/modified_dataset.csv'
df = pd.read_csv(url)
df.head()

Diabetes_binary  HighBP  HighChol  CholCheck   BMI  Smoker  Fruits  \
0              0.0     1.0       0.0        1.0  26.0     0.0     0.0   
1              0.0     1.0       1.0        1.0  26.0     1.0     1.0   
2              0.0     0.0       0.0        1.0  26.0     0.0     1.0   
3              0.0     1.0       1.0        1.0  28.0     1.0     1.0   
4              0.0     0.0       0.0        1.0  29.0     1.0     1.0   

   GenHlth  MentHlth  Sex   Age  Education  Income  
0      3.0       5.0  1.0   4.0        6.0     8.0  
1      3.0       0.0  1.0  12.0        6.0     8.0  
2      1.0       0.0  1.0  13.0        6.0     8.0  
3      3.0       0.0  1.0  11.0        6.0     8.0  
4      2.0       0.0  0.0   8.0        5.0     8.0

In [3]:
df.shape

(70692, 13)

In [4]:
# Define features set
X = df.copy()
X.drop("Diabetes_binary", axis=1, inplace=True)
X.head()

HighBP  HighChol  CholCheck   BMI  Smoker  Fruits  GenHlth  MentHlth  Sex  \
0     1.0       0.0        1.0  26.0     0.0     0.0      3.0       5.0  1.0   
1     1.0       1.0        1.0  26.0     1.0     1.0      3.0       0.0  1.0   
2     0.0       0.0        1.0  26.0     0.0     1.0      1.0       0.0  1.0   
3     1.0       1.0        1.0  28.0     1.0     1.0      3.0       0.0  1.0   
4     0.0       0.0        1.0  29.0     1.0     1.0      2.0       0.0  0.0   

    Age  Education  Income  
0   4.0        6.0     8.0  
1  12.0        6.0     8.0  
2  13.0        6.0     8.0  
3  11.0        6.0     8.0  
4   8.0        5.0     8.0

In [5]:
# Define target vector
y = df["Diabetes_binary"].values.reshape(-1, 1)
y[:5]

array([[0.],
       [0.],
       [0.],
       [0.],
       [0.]])

In [6]:
# Splitting into Train and Test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [7]:
# Creating StandardScaler instance
scaler = StandardScaler()

In [8]:
# Fitting Standard Scaller
X_scaler = scaler.fit(X_train)

In [9]:
# Scaling data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## **Fitting the Decision Tree Model**

In [10]:
# Creating the decision tree classifier instance
model = tree.DecisionTreeClassifier()

In [11]:
# Fitting the model
model = model.fit(X_train_scaled, y_train)

## **Making Predictions Using the Tree Model**


In [12]:
# Making predictions using the testing data
predictions = model.predict(X_test_scaled)

## **Model Evaluation**

In [13]:
# Calculating the confusion matrix
cm = confusion_matrix(y_test, predictions)
cm_df = pd.DataFrame(
    cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)

# Calculating the accuracy score
acc_score = accuracy_score(y_test, predictions)

In [14]:
# Displaying results
print("Confusion Matrix:")
display(cm_df)
print("-------------------------------------------")
print(f"Accuracy Score : {acc_score}")
print("-------------------------------------------")
print("Classification Report:")
print(classification_report(y_test, predictions))

Confusion Matrix:


Predicted 0  Predicted 1
Actual 0         5962         2873
Actual 1         3212         5626

-------------------------------------------
Accuracy Score : 0.6556894698127087
-------------------------------------------
Classification Report:
              precision    recall  f1-score   support

         0.0       0.65      0.67      0.66      8835
         1.0       0.66      0.64      0.65      8838

    accuracy                           0.66     17673
   macro avg       0.66      0.66      0.66     17673
weighted avg       0.66      0.66      0.66     17673



## **Analysis Question**

Would you trust in this model to predict the liklihood of individuals developing diabetes based on various health indicators?

After observing the results, the accuracy (66%) does not seem to be good enough to use this model in predicting diabetes of individuals.